jax/random.py

```diff
@@ -200,32 +207,51 @@ def _broadcast(x):
       _threefry2x32_gpu_translation_rule
 
 @jit
+@partial(vmap, in_axes=(None, 0))
 def threefry_2x32(keypair, count):
   """Apply the Threefry 2x32 hash.
 
   Args:
     keypair: a pair of 32bit unsigned integers used for the key.
-    count: an array of dtype uint32 used for the counts.
+    count: an array of dtype uint32 used for the counts. Assumed to be
+      comprised of unique pairs, so that it has shape (n, 2), with each uint32
+      pair simulating a single uint64.
 
   Returns:
     An array of dtype uint32 with the same shape as `count`.
   """
-  key1, key2 = keypair
-  if not lax.dtype(key1) == lax.dtype(key2) == lax.dtype(count) == onp.uint32:
+  if not lax.dtype(keypair) == lax.dtype(count) == onp.uint32:
     msg = "threefry_2x32 requires uint32 arguments, got {}"
-    raise TypeError(msg.format([lax.dtype(x) for x in [key1, key2, count]]))
+    raise TypeError(msg.format([lax.dtype(x) for x in [keypair, count]]))
 
-  odd_size = count.size % 2
-  if odd_size:
-    x = list(np.split(np.concatenate([count.ravel(), onp.uint32([0])]), 2))
-  else:
-    x = list(np.split(count.ravel(), 2))
+  assert np.shape(keypair) == (2,)
+  key1, key2 = keypair
 
-  x = threefry2x32_p.bind(key1, key2, x[0], x[1])
-  out = np.concatenate(x)
+  assert count.shape == (2,)
+  count_high, count_low = count
+
+  # Result of theefry2x32_p is a list
+  out = np.asarray(threefry2x32_p.bind(key1, key2, count_high, count_low))
```

Comment:
I think using concatenate or stack is more explicit (and perhaps faster) than `np.asarray`.